SimPEG/ObjectiveFunction.py

```diff
@@ -221,68 +221,41 @@ def multipliers(self, value):
-
+        f = 0.
+        for multiplier, objfct in zip(self.multipliers, self.objfcts):
+            if isinstance(multiplier, Utils.Zero) or multiplier == 0.: # don't evaluate the fct
+                pass
+            else:
+                f += multiplier * objfct(x, **kwargs)
         return f
 
     def deriv(self, x, **kwargs):
-        def eval_listDerivs(objfcts, multipliers, g):
-            for multiplier, objfct in zip(multipliers, objfcts):
-                if isinstance(multiplier, Utils.Zero) or multiplier == 0.: # don't evaluate the fct
-                    pass
-                elif isinstance(objfct, list):
-                    g += eval_listDerivs(objfct, multiplier, g)
-                else:
-                    g += multiplier * objfct.deriv(x, **kwargs)
-            return g
-
-        g = eval_listDerivs(self.objfcts, self.multipliers, Utils.Zero())
+        g = Utils.Zero()
+        for multiplier, objfct in zip(self.multipliers, self.objfcts):
+            if isinstance(multiplier, Utils.Zero) or multiplier == 0.: # don't evaluate the fct
+                pass
+            else:
+                g += multiplier * objfct.deriv(x, **kwargs)
         return g
 
     def deriv2(self, x, **kwargs):
 
-        def eval_list2Derivs(objfcts, multipliers, H):
-            for multiplier, objfct in zip(multipliers, objfcts):
-                if isinstance(multiplier, Utils.Zero) or multiplier == 0.: # don't evaluate the fct
-                    pass
-                elif isinstance(objfct, list):
-                    H = H + eval_list2Derivs(objfct, multiplier, H)
-
-                else:
-                    objfct_H = objfct.deriv2(x, **kwargs)
-                    if isinstance(objfct_H, Utils.Identity):
-                        if self.nP != '*':
-                            objfct_H = sp.eye(self.nP)  # if we need a shape
-                        elif objfct.nP != '*':
-                            objfct_H = sp.eye(objfct.nP)  # if we need a shape
-                        else:
-                            raise Exception('need to set nP on objective functions to get H')
-                    H = H + multiplier * objfct_H
-            return H
-
-        H = eval_list2Derivs(self.objfcts, self.multipliers, Utils.Zero())
+        H = Utils.Zero()
+        for multiplier, objfct in zip(self.multipliers, self.objfcts):
+            if isinstance(multiplier, Utils.Zero) or multiplier == 0.: # don't evaluate the fct
+                pass
+            else:
+                objfct_H = objfct.deriv2(x, **kwargs)
+                if isinstance(objfct_H, Utils.Identity):
+                    if self.nP != '*':
+                        objfct_H = sp.eye(self.nP)  # if we need a shape
+                    elif objfct.nP != '*':
+                        objfct_H = sp.eye(objfct.nP)  # if we need a shape
+                    else:
+                        raise Exception('need to set nP on objective functions to get H')
```

Comment:
I think lines 249 to 255 here are all side effects of the fact that `deriv2` can return a matrix if it is called without a vector, `v`  (See [`BaseRegularization`](https://github.com/simpeg/simpeg/blob/ref/objectivefunctions/SimPEG/Regularization.py#L550-L577) ). However, Considering a `BaseRegularization` class will almost always be added to a `BaseDataMisfit` like class whose implementations define a `deriv2` which **does** require a vector, `v`. (See [`l2_DataMisfit`](https://github.com/simpeg/simpeg/blob/ref/objectivefunctions/SimPEG/DataMisfit.py#L147-L163
)). Should we support only calling `deriv2` with a `v` (essentially only providing the Hessian times a vector operation)? That would simplify this code a bit :).